src/lib/viewers/ResultsViewer.ts

```diff
@@ -0,0 +1,136 @@
+			const v2PrimaryLocation = v2.getCodeLocations()[v2.getPrimaryLocationIndex()];
+			const v1File = v1PrimaryLocation.getFile() || '';
+			const v2File = v2PrimaryLocation.getFile() || '';
+			if (v1File !== v2File) {
+				return v1File.localeCompare(v2File);
+			}
+
+			// Next, compare start lines.
+			const v1StartLine = v1PrimaryLocation.getStartLine() || 0;
+			const v2StartLine = v2PrimaryLocation.getStartLine() || 0;
+			if (v1StartLine !== v2StartLine) {
+				return v1StartLine - v2StartLine;
+			}
+
+			// Finally, compare start columns.
+			const v1StartColumn = v1PrimaryLocation.getStartColumn() || 0;
+			const v2StartColumn = v2PrimaryLocation.getStartColumn() || 0;
+			return v1StartColumn - v2StartColumn;
+		});
+	}
+
+	private displaySummary(violations: Violation[]): void {
+		const violationCount = violations.length;
+		const fileCount = this.countUniqueFiles(violations);
+		this.display.displayInfo(getMessage(BundleName.ResultsViewer, 'summary.found-results', [violationCount, fileCount]));
+	}
+
+	private displayDetails(violations: Violation[]): void {
+		// If there are more results than the cutoff amount, we'll only show that many.
+		const printableViolationCount = Math.min(ResultsDetailViewer.RESULTS_CUTOFF, violations.length);
+		const omittedResultsCount = violations.length - printableViolationCount;
+		for (let i = 0; i < printableViolationCount; i++) {
+			const violation = violations[i];
+			const rule = violation.getRule();
+			const sev = rule.getSeverityLevel();
+			this.display.displayStyledHeader(
+				getMessage(
+					BundleName.ResultsViewer,
+					'summary.detail.violation-header',
+					[i + 1, rule.getName()])
+			);
+			const primaryLocation = violation.getCodeLocations()[violation.getPrimaryLocationIndex()];
+			this.display.displayStyledObject({
+				engine: rule.getEngineName(),
+				severity: `${sev.valueOf()} (${SeverityLevel[sev]})`,
+				message: violation.getMessage(),
+				location: `${primaryLocation.getFile()}:${primaryLocation.getStartLine()}:${primaryLocation.getStartColumn()}`,
+				docs: violation.getResourceUrls().join(',')
+			}, ['engine', 'severity', 'message', 'location', 'docs']);
+		}
+		if (omittedResultsCount > 0) {
+			this.display.displayInfo(getMessage(BundleName.ResultsViewer, 'summary.detail.results-truncated', [omittedResultsCount]));
+		}
+	}
+
+	private displayBreakdown(results: RunResults): void {
+		this.display.displayStyledHeader(getMessage(BundleName.ResultsViewer, 'summary.detail.breakdown.header'));
+		this.display.displayInfo(getMessage(BundleName.ResultsViewer, 'summary.detail.breakdown.total', [results.getViolationCount()]));
+		for (const sev of Object.values(SeverityLevel)) {
+			// Some of the keys will be numbers, since the enum is numerical. Skip those.
+			if (typeof sev !== 'string') {
+				continue;
+			}
+			const sevCount = results.getViolationCountOfSeverity(SeverityLevel[sev] as SeverityLevel);
```

Comment:
I really dislike how typescript uses both strings and numbers in their conversion to js for number based enums. I wish typescript should just let you iterate over the actual enumerated values and have the values contain a toString() and a toValue() method if needed. JAVA is so much better here. (end of my rant)